### PR TITLE
[TargetParser][AArch64] Believe runtime feature detection

### PR DESCRIPTION
**Pull request by workingjubilee**
In https://github.com/llvm/llvm-project/issues/90365 it was reported that TargetParser arrives at the wrong conclusion regarding what features are enabled when attempting to detect "native" features on the Raspberry Pi 4, because it (correctly) detects it as a Cortex-A72, but LLVM (incorrectly) believes all Cortex-A72s have crypto enabled. Attempt to help ourselves by allowing runtime information derived from the host to contradict whatever we believe is "true" about the architecture.

### CODE REVIEW

**Comment by workingjubilee** [comment]
I have no idea which tests are tests I should ignore the breakage of or which tests I should investigate. I have never dealt before with a tree that allows regular breakage to be checked in to its core test suite.

**Review by davemgreen — COMMENTED** [review]
Thanks for working on adding this. "crypto" is an old feature-name now, which has been split-up into aes+sha2, and the newer sha3+sm4 features in certain situations. It would technically be incorrect to turn "aes+sha2" into "crypto" if "crypto" is turned back into "sha2+aes+sha3+sm4". I think it might be aes = CAP_AES&CAP_PMULL and sha2 = CAP_SHA1&CAP_SHA2, so they don't necessarily fit into the LLVMFeatureStr map unfortunately.

Can we split this into individual sha2 and aes features? And can we add a test to TargetParserTest.cpp?

**Comment by workingjubilee** [comment]
@davemgreen That seems like blatant feature creep.

On the other hand, I am happy to incorporate such fixes to adjust the target feature schema for LLVM to make it more accurate if I am granted liberty to actually correct the inaccuracies.

**Comment by workingjubilee** [comment]
g-d forbid, this is actually done? crypto is just inferred to mean something based on aarch64 version? what.

https://github.com/llvm/llvm-project/blob/4447e255a908c4e1a2863374eaee4bc98e773c3d/llvm/lib/TargetParser/AArch64TargetParser.cpp#L184C53-L184C57

let me rephrase my previous statement: you want me to delete the crypto feature because it is incoherent nonsense, and leave it to downstreams to decide if they want to support this grab-bag of whatever, right?

**Comment by workingjubilee** [comment]
eh? but this contradicts it? er?

https://github.com/llvm/llvm-project/blob/a4b44c003e35bb6ed78af6300e576554d41f0368/llvm/lib/Target/AArch64/AArch64Features.td#L134-L144

hm, so the part that does the inference predates the other, but only by a few months, this year...

**Comment by workingjubilee** [comment]
...and why is crypto detection behind the aarch64 ifdef? 32-bit Arm CPUs can support aes and sha2, can't they? I mean maybe they usually won't, but it doesn't seem correct to refuse to detect it if the kernel indicates otherwise.

**Comment by davemgreen** [comment]
Hi - That is true, it has always been like that so shouldn't be the job of this patch to try and fix it. If you can add some testing though (hopefully to TargetParserTest.cpp) this should be OK.

**Comment by workingjubilee** [comment]
I don't see how native feature detection is tested in TargetParserTest.cpp? Is it just... not?

We'd need to mock the host detection functions called in order to do so, right?

**Comment by davemgreen** [comment]
Oh I was meaning in llvm/unittests/TargetParser/Host.cpp. It has tests for parsing things that look like cpuinfo.

**Comment by workingjubilee** [comment]
> Oh I was meaning in llvm/unittests/TargetParser/Host.cpp. It has tests for parsing things that look like cpuinfo.

Sure, but it doesn't have any tests for `sys::getHostCPUFeatures`, because there are none, because the entire schematic for looking up host features is "do something target-dependent at runtime", and thus is currently untestable, unless someone alters the signature for that function for every single platform...

...and considering my incredibly meager C++ knowledge, it seems _less_ effort to simply fix the definitions of the Armv8-A Cortex targets.

**Comment by workingjubilee** [comment]
This only becomes unit-testable, even for just AArch64, with a reasonable amount of effort (as opposed to the unreasonable amount I just cited) if it is acceptable to kill off the ifdefs and then expose the internal string -> StringMap parser.

But it may be a better idea to instead just switch to using HWCAP and HWCAP2?

**Review by davemgreen — COMMENTED** [review]
OK. We would usually want some form of testing for a change like this, but it might be alright so long as we verify it is working manually. As you can tell this bit of code isn't super well maintained, and could do with a number of improvements.

Are you sure this function gets called for -mcpu=native, or are you using this from somewhere that isn't clang?

**Comment by workingjubilee** [comment]
Ehm, yes? https://github.com/llvm/llvm-project/blob/9f44d5d9d0903adaa9deb35d33056202e5030cb3/clang/lib/Driver/ToolChains/Arch/ARM.cpp#L592-L605

We primarily use this: https://github.com/llvm/llvm-project/blob/9f44d5d9d0903adaa9deb35d33056202e5030cb3/llvm/include/llvm-c/TargetMachine.h#L231-L233

**Comment by davemgreen** [comment]
OK. I think this might be used in clang for ARM but not AArch64 at the moment. If you can change the variable names, and you have tested this in your use-case (rust?) then this looks OK to me.

**Comment by ElvinaYakubova** [comment]
@workingjubilee Hi! Do you plan to continue working on this?

**Comment by workingjubilee** [comment]
Did anyone get the license plate of that runaway thought?

**Review by davemgreen — APPROVED** [review]
This looks sensible to me. Thanks for making the change. I don't believe it gets used from clang yet, hopefully #97749 can improve upon that.

LGTM

**Comment by workingjubilee** [comment]
Thank you for the review! Sorry about getting brain-flattened.
